src/pil/ir/powdr_pil.rs

```diff
@@ -0,0 +1,29 @@
+use crate::{ast::query::Queriable, poly::Expr, util::UUID};
+use std::collections::HashMap;
+extern crate regex;
+
+// PIL circuit IR
+pub struct PilCircuit<F> {
+    // circuit level
+    pub circuit_name: String,
+    pub num_steps: usize,
+
+    // circuit level - col witness
+    pub col_witness: Vec<UUID>, // internal signal NOT dedupped
+
+    // circuit level - col fixed
+    pub fixed_signals: HashMap<UUID, Vec<F>>, // fixed signal UUID -> fixed assignments vector
+    pub step_types_instantiations: HashMap<UUID, Vec<usize>>, /* step type UUID -> step
```

Comment:
@qwang98 there is no steps in PIL, the PIL shouldn't have steps. The fact that this is here might mean that there is a translation that you are doing in the backend that should be done in the compiler

Reply:
Agreed that PIL shouldn't have steps, but we do need to express steps and step instantiations in PIL through fixed signals. For example: https://github.com/privacy-scaling-explorations/chiquito/blob/8803c3a47793a37ea7d039c63e68434d87d7d472/examples/mimc7.pil#L7-L10

I removed `step_types_instantiations` from PIL IR and converted all of them to `fixed_signals` in compiler instead.

Note that we still need `first_step` and `last_step` fields in PIL IR, although they are not PIL concepts. We need them to enforce things like `pragma_first_step` and `pragma_num_step` in PIL, which are enforced through constraints like the following example: https://github.com/privacy-scaling-explorations/chiquito/blob/8803c3a47793a37ea7d039c63e68434d87d7d472/examples/mimc7.pil#L43-L46

To your point, we could technically convert `first_step` and `last_step` to `constraints` in PIL IR, but that'll require modifying the annotation HashMap and the layers that use this data structure and can be quite error prone, so I elected to keep `first_step` and `last_step` in PIL IR.